Add DNS service discovery

emrob opened the pull request:
Add DNS service discovery using [dnsdock](https://github.com/tonistiigi/dnsdock). When `docker-compose up` is run, the dnsdock container automatically starts and allows services to discover each other on the .local.edx.org domain. Dnsdock requires that the DNS listening port is bound to the docker0 interface, so the docker daemon must be started with options `--bip=172.17.42.1/24` and `--dns=172.17.42.1` to use dnsdock.

edx-webhook commented:
Thanks for the pull request, @emrob! I've created [OSPR-871](https://openedx.atlassian.net/browse/OSPR-871) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).
We can't start reviewing your pull request until you've   submitted a [signed contributor agreement](http://open.edx.org/sites/default/files/wysiwyg/individual-contributor-agreement.pdf)   or indicated your institutional affiliation   and   added yourself to the [AUTHORS](https://github.com/edx/configuration/blob/master/AUTHORS) file . Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

feanil commented on docker/compose/docker-compose.yml:
```diff
@@ -1,40 +1,67 @@
+    - /var/run/docker.sock:/var/run/docker.sock
+  ports:
+    - 172.17.42.1:53:53/udp
+  command: -domain="local.edx.org"
```
Is this right, shouldn't there be a command before the parameter?

emrob replied:
It looks a little odd, but that should be correct. Docker images have both an Entrypoint and a Cmd. The Entrypoint is the executable that runs when a container starts and the Cmd is essentially passed to the Entrypoint as an argument (see [here](https://stackoverflow.com/questions/21553353/what-is-the-difference-between-cmd-and-entrypoint-in-a-dockerfile)). Dnsdock's Entrypoint is `/go/bin/dnsdock`, so when the container starts, we're running `/go/bin/dnsdock -domain="local.edx.org"`.

sarina commented:
@emrob could you rebase your branch?

@feanil FYI we have Morgan's contributor info, this is ready for review.

feanil commented:
@emrob looks like this needs a rebase, can you do that?

@e0d any objections to merging this into the docker-hack branch?

emrob commented:
@feanil @e0d I rebased and fixed some of the links to make them consistent with how dnsdock [does matching](https://github.com/tonistiigi/dnsdock#user-content-usage). Things like `lms-mysql.local.edx.org` won't match `mysql.local.edx.org`, so I replaced the dashes with dots. Is there a specific reason that we're prepending the DNS name with the service's name? It might make as much sense to leave the service off and just say `mysql.local.edx.org`.

e0d commented:
@emrob I think we decided this could be closed now in favor of using docker networking

emrob commented:
@e0d Agreed
